2022/Q3/artifacts/PoS-tla/Staking.tla

```diff
@@ -113,104 +153,245 @@ Delegate(sender, validator, amount) ==
     /\ lastTx' = [ id |-> nextTxId, tag |-> "delegate", fail |-> fail,
                    sender |-> sender, toAddr |-> validator, value |-> amount ]
     /\ failed' = (fail \/ failed)
+    /\ nextTxId' = nextTxId + 1
+    /\ txCounter' = txCounter + 1
+    /\ UNCHANGED <<epoch, unbonded, idUnbonds, slashes, idSlashes, enqueuedSlashes, frozenValidators>>
     /\  IF fail
         THEN
-          UNCHANGED <<balanceOf, unbonded, delegated, bonded>>
+          UNCHANGED <<balanceOf, posAccount, bonded, idBonds, totalDeltas>>
         ELSE
           \* transaction succeeds
           \* update the balance of 'sender'
           /\ balanceOf' = [ balanceOf EXCEPT ![sender] = @ - amount]
-          /\ delegated' = [ n \in 1..UnbondingLength, user \in UserAddrs, val \in ValidatorAddrs |-> 
-                     IF n >= PipelineLength /\ user = sender /\ val = validator
-                     THEN delegated[n, user, val] + amount
-                     ELSE delegated[n, user, val]
-                    ]
-          /\ bonded' = [ n \in 1..UnbondingLength, user \in ValidatorAddrs |-> 
-                     IF n >= PipelineLength /\ user = sender
-                     THEN bonded[n, user] + amount
-                     ELSE bonded[n, user]
-                    ]
-          /\ UNCHANGED unbonded
-
-
-\* unbond `amount` tokens from a validator
+          /\ posAccount' = posAccount + amount
+          /\ bonded' = [ bonded EXCEPT ![sender, validator] = @ \union {[ id |-> idBonds,
+                                                                          epoch |-> epoch + PipelineLength,
+                                                                          amount |-> amount ]}]
+          /\ idBonds' = idBonds + 1
+          \* updates totalDeltas from PipelineLength to UnbondingLength 
+          /\ totalDeltas' = [ n \in 0..2*UnbondingLength, val \in ValidatorAddrs |-> totalDeltas[n, val] + 
+                              IF n >= UnbondingLength + PipelineLength /\ val = validator
+                              THEN amount
+                              ELSE 0
+                            ]
+
+\* Follows the specification: it counts all bonds, even those that have not been materialized yet.
+\* PAY ATTENTION TO IT
+\* @type: (ADDR, ADDR) => Int;
+TotalBonds(sender, validator) == LET 
+                                 \* @type: (Int, BOND) => Int;
+                                 F(total, bond) == total + bond.amount
+                                 IN ApaFoldSet(F, 0, bonded[sender, validator])
+
+\* @type: (Set(BOND), Int, Int) => [ remaining: Int, set: Set(UNBOND), id: Int ];
+ComputedUnbonds(setBonds, totalAmount, e) == LET 
+                                         \* @type: ([ remaining: Int, set: Set(UNBOND), id: Int ], BOND) => [ remaining: Int, set: Set(UNBOND), id: Int ];
+                                         F(record, bond) == 
```

Comment:
Would it be clearer to rename "bond" to "unbond" here?

Reply:
Do you mean the bond in `F(record, bond)`?

Reply:
Yes, the parameter's name

Reply:
Not sure I follow, it is a bond and not an unbond.

Reply:
Yeah, I get it now (thanks for offline explanation)